### ISSUE TITLE
Improve browsing experience for documentation

### ISSUE THREAD

**Issue by kevinsimper** [hapijs/hapijs.com#200]
Sometimes it can be really hard to see where properties is belonging to, for example the `server.route` has a lot of nested configuration, and sometimes you miss that it has to nested a chain of properties.

I have missed that `payload.parse` should be inside the `config` object and not as a property inside the `server.route`.

Accordion
https://jqueryui.com/accordion/#default

**Comment by mik01aj** [comment]
+1 +1 +1. The documentation is very hard to read.

Hapi has a lots of alternative-options-in-configs-in-callbacks-to-I-dont-know-what and this becomes overwhelming when the documentation doesn't have a good visual structure. Here's an example of handler options to the `route` method's `handler` option (afaik normally there's a function there, but you can also have an object):

![screenshot 2015-08-19 09 40 23](https://cloud.githubusercontent.com/assets/3027415/9351088/9e13bf56-4656-11e5-8f15-c66d0e983928.png)

You have your headers in almost the same font face as everything else. The vertical spacing is so big, and the list indent so small that you hardly see which level of nesting you're looking at. (Actually, four levels of nested lists are usually not a good idea, and your docs contain lots of places with 4 levels, and even some with 5 levels).

In short, I mean: some good CSS would help as well. Expanding/collapsing, yes, it's a good idea, but some better styling is needed too. And this should be actually much easier.

It is sooooo easy to get lost.

**Comment by kevinsimper** [comment]
@nlf How can we make this happen?

**Comment by AdriVanHoudt** [comment]
since `app.js` starts with `/* these methods borrowed from youmightnotneedjquery.com */` I assume we need to find a solution without jquery :P

**Comment by mtharrison** [comment]
I agree that some simple styling would help too. Some subtle borders on the `<ul>`s might improve things:

![screen shot 2015-08-19 at 10 13 12](https://cloud.githubusercontent.com/assets/916064/9352727/6065a9bc-465b-11e5-90ba-3ea34c0c78f5.png)

**Comment by Marsup** [comment]
Putting "help wanted" suggests that @nlf is waiting for a PR.

**Comment by AdriVanHoudt** [comment]
@mtharrison nice idea but looks ugly ^^

**Comment by kevinsimper** [comment]
@Marsup But with no information on what they expect and what they think, it is hard to know if the time spent is worth it.

**Comment by Marsup** [comment]
That debate is already a good start, if you all converge on a solution I bet it'll take.

**Comment by dj** [comment]
+1

My team uses hapi and I can report that we all find the documentation to be really difficult to read. I think a good solution would change the documentation from a gigantic unordered list to a more structured documentation format. I'll try to think of a configuration heavy framework like hapi that might serve as an example.

I think that the styling proposed by @mtharrison would be super helpful in the short term, but ultimately the documentation needs to be improved in a more fundamental way. I think if anyone ( @AdriVanHoudt ) thinks its ugly, they should suggest alternative styles but saying 'it's ugly' on it's own is not useful feedback.

I'm going to chew on this for a bit, would love to hear suggestions from other users as well as examples of similar framework documentation that you like.

**Comment by AdriVanHoudt** [comment]
@dj I agree my feedback is not super constructive but I am also not a designer ;) I can ask our designer for some ideas if he is willing though

**Comment by mik01aj** [comment]
@dj http://lesscss.org/features/ is also one long page, but thanks to the navigation on the left it's easy to understand where you are and navigate. I don't think that Hapi docs need some massive restructuring to be readable. And I did give some constructive feedback by pointing out some stuff that is easily fixable:

> You have your headers in almost the same font face as everything else. The vertical spacing is so big, and the list indent so small that you hardly see which level of nesting you're looking at.

Plus the side navigation. Imho it's the most important part here.

**Comment by dj** [comment]
@mik01aj to be honest, I think I skipped over your comment the first time I read this thread (my bad), but I agree with your suggestions.

I think the LESS side nav would go a long way to helping the reader keep track of where they are in long lists of configuration options and agree that changing the typography to better emphasize the hierarchy would go a long way.

I also agree that a 'massive restructuring' isn't needed to increase readability, and my initially comment might have been a little bit too zealous. I do think that the giant `<ul>` needs to be broken up into smaller lists with headers in between. I think we also agree that better typography would do a lot more than adding expanding/collapsing.

**Comment by AdriVanHoudt** [comment]
yeah I tried just turning of the font definition in css and the default more bolder, less white space font already does a lot, the sidenav that scrolls with you would also do a lot +1 on both

**Comment by kevinsimper** [comment]
I raised the margin-left from 20px to 30px and then made a `border-left: 1px dashed rgba(0, 0, 0, 0.4);`

It is easier now to follow the lines all way to the top.

![image](https://cloud.githubusercontent.com/assets/1126497/9440202/92957d02-4a6d-11e5-8995-30eb5366aee6.png)

**Comment by kevinsimper** [comment]
@AdriVanHoudt I just tried `git clone` and after running the commands in the readme, the api page is empty.

**Comment by AdriVanHoudt** [comment]
did you update the github token? also it takes a while the first time

**Comment by nlf** [comment]
i originally had the navigation following the user's scroll position, i forget why it was removed but i think it was when i started having github render the markdown to resolve some weird parsing problems.

that said, i think it would be a good idea to put it back. avoiding jquery would be nice, but it's definitely not necessary. mostly i was just trying to keep things fairly lightweight, but that's an arbitrary restriction that can certainly be ignored.

**Comment by theotow** [comment]
To push this a little, i also had trouble getting through those huge amounts of text to find a specific setting.  

Currently: 
the docs are separated in text that describes object keys verbose +  example code snippet 

Idea:

```
{
  cache: 'sets up server-side caching' // cross reference cache Strategy Object
  ... more keys here
  load: {
    sampleInterval: 0 // the frequency of sampling in milliseconds
  }, 
  mime: 'options passed to the mimos module' // cross reference mimos config object 
  plugins: {
    'pluginname': 'configuration' // cross reference Configuration of Plugin 
  },
  useDomains: false // protect against exceptions thrown in handlers
}
```
This would involve 

1. Showing the full config object prepopulated with every possible key and default value + very short description or cross reference to corresponding object 

2. toggle (clickable) at each key in the object to read more about the option for example 'useDomains' => if false, will not use node domains to protect against exceptions thrown in handlers and other external code. Defaults to true.

3. root toggle at each section to expand all toggles in these section ( to maintain searchability ) 

4. still keep the example code below


Technical Side:

I could image we can use a JSON prettifier and have a hover state for each line in the which will signalize right or left that this key is expandable with more information or not.

UX:

Just checked on my macbook, the hapi doc is comparable with a ~ 169 pages book, calculated by the height of the page / browser window height. I think this could be dramatically reduced.

**Comment by johnmanko** [comment]
It would be better to add a sidebar nav similar to Bootstrap's docs: 
https://getbootstrap.com/components/#nav

**Comment by sirgallifrey** [comment]
what if we make it into something similar to the stripe docs? https://stripe.com/docs/api#intro

We could use [Slate](https://github.com/lord/slate) changing it's ruby build to javascript, or something else similar

**Comment by sirgallifrey** [comment]
Here is a very good exemple of documentation: [https://learn.getgrav.org/basics/grav-configuration](https://learn.getgrav.org/basics/grav-configuration)

Instead os nesting configurations items, there is a table and you have the config "paht" on each row, like the row **Author: Name:**

If we use this on hapi docs we will get rid of the 5 levels of nesting and I think ctrl+f could be easier

**Comment by AdriVanHoudt** [comment]
@sirgallifrey you had a picture of how the docs could look with slate on gitter, could you share that here as well, it will give everyone a better idea I think (also it looked quite good ^^)

**Comment by sirgallifrey** [comment]
![hpaidocs slate](https://cloud.githubusercontent.com/assets/2331645/20859443/99fac360-b946-11e6-8543-67c28f12c08c.png)

**Comment by thebergamo** [comment]
I like the approach of the @sirgallifrey, maybe if we separate the docs from the main website and use an alteranative approach like the sequelize[1] docs or some specific stylish of documentation will be improve the reads; 

[1] Sequelize documentation example: http://docs.sequelizejs.com/en/v3/api/sequelize/#new-sequelizeuri-options

**Comment by AdriVanHoudt** [comment]
I don't think we should separate the docs from the website. 
And personally I think slate looks better than read the docs (the thing sequalize uses)

**Comment by thebergamo** [comment]
I like read the docs better too, but the idea is just an example. But when the docs have a lots of methods e properties, maybe separate or embed a read the docs inside will be a good approach.

**Comment by AdriVanHoudt** [comment]
Yeah I think we definitely want to keep the docs on the website

**Comment by sirgallifrey** [comment]
@thebergamo when you say separate the docs and the website you are talking about the two of them having different visual styles or making two websites in fact? In any case I don't think that separating the docs from the website will fix the docs... 
I have played a little more with Slate and I can say it's not a "silver bullet", Slate will help fixing some of the problems but will bring others to the table...

**Comment by nlf** [comment]
something to keep in mind - a lot of effort went into making sure that the docs render as closely to how github does it as possible. github does a lot of things differently with markdown that we need to be cognizant of since that's what they target (and should!)

**Comment by AdriVanHoudt** [comment]
@nlf why does it need to look like the md on github?

**Comment by nlf** [comment]
because that's where the markdown is written and what it targets. i know on occasion the docs are written using github's preview feature and i'm not about to ask any maintainers to use anything beyond that. github has a lot of features that it adds to markdown, along with a lot of things that it does differently (rendering headers as anchors, for example).

i don't mind if it *looks* different. like if we were to render certain things side-by-side like slate, but making sure that the *features* of github style markdown are present and functional is definitely critical.

**Comment by AdriVanHoudt** [comment]
I totally agree that the source should remain markdown and that the website should just enhance it. So yeah something like slate or what we have now (but clearer) is what we need to go for

**Comment by thebergamo** [comment]
In this case, organizing better the information will be suits fine.

**Comment by evanperriello** [comment]
Just played around with this a bit as an interim solution. I integrated the borders @mtharrison suggested, changed the spacing and the font sizing, and nixed the bullet points in favor of first-line indentation to better highlight the important terms that start off each list item. This isn't a full restructure, which might be needed in future, but in the interim it might make things a bit more user-friendly.
![interimdocstyling](https://cloud.githubusercontent.com/assets/17691095/23371826/6244f2f8-fce7-11e6-8a50-1b6efd1180d2.JPG)

**Comment by evanperriello** [comment]
There are two small changes that I think might help the secondary nav on this page:
1. A first-line outdent for each li might distinguish them better from second or third lines of the preceding li, as below.
![indentnav](https://cloud.githubusercontent.com/assets/17691095/23411157/7ce13c58-fd9f-11e6-95aa-7e679e1445ed.JPG)

2. Right now, every item includes its full label (i.e., "server.auth.scheme" and "server.auth.strategy"). I wonder what you all would think about cutting out "server" as follows and relying on the nesting to make that apparent (see example below). That way, folks could see the property/method they're looking for more quickly.

Server
&nbsp;&nbsp;&nbsp;&nbsp;.auth.api
&nbsp;&nbsp;&nbsp;&nbsp;.auth.default(options)
&nbsp;&nbsp;&nbsp;&nbsp;.auth.scheme(name, scheme)

**Comment by fe-artisan** [comment]
+1.

what if we make it into something similar to the webpack config docs?

https://webpack.js.org/configuration/

**Comment by AdriVanHoudt** [comment]
I like their gitter button

**Comment by apipkin** [comment]
I would be happy to work on this but I am unable to get a local version of the hapi site running. When going to `/api` I get a 500 error page. The server seems to be responsive but there is no content showing. If someone can help me out with this, I'll see what I can put together to help the api doc experience.

**Comment by hueniverse** [comment]
Closing. If people have specific new ideas, please open issues. This 3 years long discussion hasn't been helpful.
